operator/pkg/ciliumidentity/controller_test.go

```diff
@@ -197,3 +199,244 @@ func verifyCIDUsageInCES(ctx context.Context, fakeClient *k8sClient.FakeClientse
+				// https://github.com/cilium/cilium/issues/34740
+				{cid1.SecurityLabels, cid2.SecurityLabels},
+			},
+		},
+		{
+			name: "reuse_cid",
+			podActions: []podAction{
+				{"create", pod1, nil},
+				{"delete", pod1, nil},
+				{"create", pod1, nil},
+			},
+			expectedInCIDStore: [][]map[string]string{
+				{cid1.SecurityLabels},
+				{cid1.SecurityLabels},
+				{cid1.SecurityLabels},
+			},
+		},
+		{
+			name: "update_cid",
+			podActions: []podAction{
+				{"create", pod1, nil},
+				{},
+			},
+			cidUpdateList: []*cidUpdate{
+				nil,
+				{cid1.SecurityLabels, cid2},
+			},
+			expectedInCIDStore: [][]map[string]string{
+				{cid1.SecurityLabels},
+				{cid2.SecurityLabels},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			cid, _, fakeClient, _, h := initHiveTest(true)
+			ctx := context.Background()
+			tlog := hivetest.Logger(t)
+			if err := h.Start(tlog, ctx); err != nil {
+				t.Fatalf("starting hive encountered an error: %s", err)
+			}
+			defer func() {
+				if err := h.Stop(tlog, ctx); err != nil {
+					t.Fatalf("stopping hive encountered an error: %v", err)
+				}
+			}()
+
+			if _, err := fakeClient.Slim().CoreV1().Namespaces().Create(ctx, ns1, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("create namespace: %v", err)
+			}
+
+			cidStore, _ := (*cid).Store(ctx)
+
+			// Test starts here.
+			// Run podActions in the sequence defined in the test case.
+			// Verify CID state after each podAction.
+			for actionIdx, podAction := range tc.podActions {
+				pod := podAction.podObj
+				switch podAction.action {
+				case "create":
+					if _, err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{}); err != nil {
+						t.Fatalf("create pod: %v", err)
+					}
+				case "update":
+					if _, err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Update(ctx, pod, metav1.UpdateOptions{}); err != nil {
+						t.Fatalf("update pod: %v", err)
+					}
+				case "delete":
+					if err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{}); err != nil {
+						t.Fatalf("delete pod: %v", err)
+					}
+				}
+
+				if len(tc.cidUpdateList) > 0 && tc.cidUpdateList[actionIdx] != nil {
+					cidList := cidStore.List()
+					if len(cidList) == 0 {
+						t.Fatal("listed CIDs for update = 0, want >= 1")
+					}
+
+					cidUpdateRef := tc.cidUpdateList[actionIdx]
+					var cidToUpdate *capi_v2.CiliumIdentity
+					// Get from CID store the CID that needs to be updated.
+					for _, cid := range cidList {
+						if cmp.Equal(cid.SecurityLabels, cidUpdateRef.origSecLabels) {
+							cidToUpdate = cid
+							break
+						}
+					}
+
+					if cidToUpdate == nil {
+						t.Fatalf("CID for update not found in CID store, labels: %v", cidUpdateRef.updatedCID.SecurityLabels)
+					}
+
+					cidToUpdate.Labels = cidUpdateRef.updatedCID.Labels
```

Comment:
Here we are modifying a reference to an element in the Store that is read-only. We should first make a deep copy.

https://github.com/cilium/cilium/blob/6d114bad0556b73a01d0eed0ca991f958e9340a1/pkg/k8s/resource/resource.go#L92-L96

More importantly, I realized we are doing the same in the controller :grimacing:  :

https://github.com/cilium/cilium/blob/6d114bad0556b73a01d0eed0ca991f958e9340a1/operator/pkg/ciliumidentity/reconciler.go#L228-L229

Reply:
Nice catch :) I've drafted https://github.com/cilium/cilium/pull/34805. Checked the other usages in reconciler.go and there is no other place where we do this. Though this in practice should not create a bug as we immediately call the clientset to update it. But indeed there could be cases in which this goes wrong.

Reply:
I decoupled the update test and applied the pattern of copying the store object before updating it.